Change the API and add documentaion

davidt99 opened the pull request:
I mainly moved stuff but the major change that I made is using string as a return value instead of `FuzzyHash` struct.
The documentation is almost like the c documentation.

davidt99 commented:
The benchmark result:
```
BenchmarkDistance-8        9437          9083          -3.75%
BenchmarkRollingHash-8     3.20          3.10          -3.12%
BenchmarkSumHash-8         35.2          33.2          -5.68%
BenchmarkBlockSize-8       6.85          6.58          -3.94%
BenchmarkProcessByte-8     31.2          30.4          -2.56%
```

glaslos commented on ssdeep.go:
```diff
@@ -137,66 +112,142 @@ type fuzzyReader interface {
+	buffer []byte
+}
+
+func (d *digest) Write(p []byte) (n int, err error) {
```
My concerns https://github.com/glaslos/ssdeep/issues/11#issuecomment-353777252 are regarding that the `Write()` interface is usually used to update the digest and discard the written data. In your implementation you are collecting data in memory on `Write` and then calculate the hash on `Sum`. As said, not really something we can fix as long as we don't know the final amount of data to be written.

davidt99 replied:
You are right, but ssdeep is not "classic" hash, so there is no much we can do. I'll be more clear in the documentation.

glaslos replied:
Sounds good and I think the added information should be good for now. We can also add a note in the README later.

davidt99 replied:
It's a good idea. Can you update the README after the release with this caveat?

davidt99 commented on ssdeep.go:
```diff
@@ -137,66 +112,142 @@ type fuzzyReader interface {
+// FuzzyFilename will opens, reads, and hashes the contents of the file 'filename'.
+// It is the caller's responsibility to append the filename to the result after computation.
+// Returns an error when the file doesn't exist or ssdeep could not be computed on the file.
+func FuzzyFilename(filename string) (string, error) {
```
I had few problems with `FuzzyHash`, mainly, that it couldn't be constructed outside of the library, so if I had a ssdeep hash string, I wasn't able to compare it to other hash.
Looking at few other implementations (c, python, ruby), all of them are returning just the string, and parsing the string when comparing two hashes.

davidt99 commented:
@glaslos I've changed some of what you requested.
I deleted `app/ssdeep.go` because I thought it was an example, my mistake, I restored it.
Tell me if there are any more issues.

glaslos commented on ssdeep.go:
```diff
@@ -137,66 +112,142 @@ type fuzzyReader interface {
 
-func (sdeep *SSDEEP) process(r *bufio.Reader) {
-	sdeep.newRollingState()
+func (state *ssdeepState) process(r *bufio.Reader) {
```
@davidt99 what's your opinion on this?

glaslos commented on ssdeep.go:
```diff
@@ -137,66 +112,142 @@ type fuzzyReader interface {
+// If an error occurs, the file pointer's value is undefined.
+// It is the callers's responsibility to append the filename to the result after computation.
+// Returns an error when ssdeep could not be computed on the file.
+func FuzzyFile(f *os.File) (string, error) {
```
@davidt99 did you think about this? IMO  we should try to finalize the API before we do a 1.0 release.

davidt99 replied:
Didn't notice that. It's a good idea to expose `FuzzyReader`, but one concern I have is it might not be intuitive to the caller to pass `*os.File` to a function that have `FuzzyReader` as a parameter. What do you say about adding new API that accepts `FuzzyReader`?

glaslos replied:
Like this one? https://github.com/intezer/ssdeep/blob/19cb84b1d6ec7f1b00bcf4a0d091a2d7328b4489/ssdeep.go#L124
We should also make a distinction between the type fuzzyReader and the method fuzzyReader

davidt99 replied:
This one exactly, I'll rename the struct then

glaslos commented:
Two more discussion items but we are pretty much good to merge.

davidt99 commented:
I think I changed my mind regarding `Hash`, ssdeep doesn't fit it. `BlockSize` and `Size` is unreliable in our case. I'm considering deleting it, @glaslos what do you think?

glaslos commented:
I agree, most of the benefits of `Hash` don't apply in our case. We could make it a sub package in case someone needs it? But I see little use. If you agree, feel free to remove it.

davidt99 commented:
So I'll delete it

davidt99 commented:
The new benchmark:
```
BenchmarkDistance-8        8710          8719          +0.10%
BenchmarkRollingHash-8     3.00          2.88          -4.00%
BenchmarkSumHash-8         31.1          31.5          +1.29%
BenchmarkBlockSize-8       6.43          6.43          +0.00%
BenchmarkProcessByte-8     28.0          28.3          +1.07%
```

glaslos commented:
Awesome, I think this was a really good process with a great result! Thanks for all the work.
